server/src/main/java/org/opensearch/action/admin/cluster/remotestore/RemoteStoreService.java

```diff
@@ -0,0 +1,185 @@
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.metadata.Metadata;
+import org.opensearch.cluster.metadata.RepositoriesMetadata;
+import org.opensearch.cluster.metadata.RepositoryMetadata;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.repositories.RepositoriesService;
+import org.opensearch.repositories.Repository;
+import org.opensearch.repositories.RepositoryVerificationException;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+/**
+ * Contains all the method needed for a remote store node lifecycle.
+ */
+public class RemoteStoreService {
+
+    private static final Logger logger = LogManager.getLogger(RemoteStoreService.class);
+    private final Supplier<RepositoriesService> repositoriesService;
+    private final ThreadPool threadPool;
+    public static final Setting<String> REMOTE_STORE_COMPATIBILITY_MODE_SETTING = Setting.simpleString(
+        "remote_store.compatibility_mode",
+        CompatibilityMode.STRICT.value,
+        CompatibilityMode::validate,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    /**
+     * Node join compatibility mode introduced with remote store.
+     *
+     * @opensearch.internal
+     */
+    public enum CompatibilityMode {
+        STRICT("strict"),
+        ALLOW_MIX("allow_mix");
+
+        public static CompatibilityMode validate(String compatibilityMode) {
+            try {
+                return CompatibilityMode.valueOf(compatibilityMode.toUpperCase(Locale.ROOT));
+            } catch (IllegalArgumentException e) {
+                throw new IllegalArgumentException(
+                    "["
+                        + compatibilityMode
+                        + "] compatibility mode is not supported. "
+                        + "supported modes are ["
+                        + CompatibilityMode.values().toString()
+                        + "]"
+                );
+            }
+        }
+
+        public final String value;
+
+        CompatibilityMode(String value) {
+            this.value = value;
+        }
+    }
+
+    public RemoteStoreService(Supplier<RepositoriesService> repositoriesService, ThreadPool threadPool) {
+        this.repositoriesService = repositoriesService;
+        this.threadPool = threadPool;
+    }
+
+    /**
+     * Performs repository verification during node startup post its creation by invoking verify method against
+     * repository mentioned. This verification will happen on a local node to validate if the node is able to connect
+     * to the repository.
+     */
+    public void verifyRepository(List<Repository> repositories, DiscoveryNode localNode) {
```

Comment:
Can we move this to repository service. Is this a code duplication?

Reply:
@psychbot : have you explored the option to add remote repo information at the node startup when initial cluster state is generated in GatewayMetaState 
https://github.com/opensearch-project/OpenSearch/blob/8cfde6ca4031047b4f7e5ecf706e434e1c535567/server/src/main/java/org/opensearch/gateway/GatewayMetaState.java#L216

This can trigger repo verification as well and need not wait for election to finish to perform the verification

Reply:
During the mixed mode, the enforcement would be disabled. But the cluster manager nodes should always have remote repository right? Otherwise how will the remote store node function? Also at that time you still need to ensure same repo is present in all the remote store node attributes.

Reply:
We should look at decoupling repo verification from cluster state addition, one way we are doing this is parsing the yml  and verifying repo on node startup and then failing bootstrap if yml configurations aren't accurate to let the node connect to repo(fail-fast). During cluster metadata bootstrap we just look at the verified repo metadata and register it in the cluster state provided compatibility checks across nodes holds true